File: 2021-remove-all-occurrences-of-a-substring/remove-all-occurrences-of-a-substring.java
class Solution {
    public String removeOccurrences(String s, String part) {
        StringBuilder curr = new StringBuilder();
        int n = part.length();
        char en = part.charAt(n - 1);

        for (int i = 0; i < s.length(); i++) {
            char ch = s.charAt(i);
            curr.append(ch);

            if (ch == en && curr.length() >= n) {
                String tail = curr.substring(curr.length() - n, curr.length());
                if (tail.equals(part)) {
                    curr.delete(curr.length() - n, curr.length());
                }
            }
        }

        return curr.toString();
    }
}

File: 2021-remove-all-occurrences-of-a-substring/remove-all-occurrences-of-a-substring.py
class Solution:
    def removeOccurrences(self, s: str, part: str) -> str:
        curr = [] # empty list
        n = len(part) # len of pattern
        end = part[-1] # end char of pattern

        for i in s:
            curr.append(i) # appending character in empty list
            if i == end and len(curr) >= n: # while iterating if i is the end ch of that pattern and the len of the curr list is >= n that means its possible substr exists
                if "".join(curr[-n:]) == part: # traverse backwards in curr to see if substr matches
                    del curr[-n:] # if yes the delete it from curr
        return "".join(curr) # return whatever's left